File: src/workspaceforge/util/prompt.ts
import * as readline from "node:readline/promises";

/** Ask a free-form question and return the trimmed answer. */
export async function ask(question: string): Promise<string> {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  try {
    const answer = await rl.question(question);
    return answer.trim();
  } finally {
    rl.close();
  }
}

/** Ask a yes/no question. Defaults to `defaultYes` if the user just presses enter. */
export async function askYesNo(question: string, defaultYes = true): Promise<boolean> {
  const suffix = defaultYes ? "[Y/n]" : "[y/N]";
  const answer = (await ask(`${question} ${suffix} `)).toLowerCase();
  if (answer === "") {
    return defaultYes;
  }
  return answer === "y" || answer === "yes";
}

export type RepairResetAbortChoice = "repair" | "reset" | "abort";

/** Ask the repair / reset / abort question used by partial-state handling. */
export async function askRepairResetAbort(): Promise<RepairResetAbortChoice> {
  const answer = (
    await ask("Choose an option: [r]epair / reset ([w]ipe and start clean) / [a]bort: ")
  ).toLowerCase();
  if (answer.startsWith("r") && !answer.startsWith("re")) {
    return "repair";
  }
  if (answer === "repair") {
    return "repair";
  }
  if (answer.startsWith("w") || answer === "reset") {
    return "reset";
  }
  return "abort";
}

export function isInteractiveTerminal(): boolean {
  return Boolean(process.stdin.isTTY) && Boolean(process.stdout.isTTY);
}

File: python/src/agenticworkspace/util/prompt.py
"""
Interactive terminal prompts used by the `init` command's partial-state
repair/reset/abort flow. Ported from src/agenticworkspace/util/prompt.ts.
"""
from __future__ import annotations

import sys
from typing import Literal

RepairResetAbortChoice = Literal["repair", "reset", "abort"]


def ask(question: str) -> str:
    """Ask a free-form question and return the trimmed answer."""
    answer = input(question)
    return answer.strip()


def ask_yes_no(question: str, default_yes: bool = True) -> bool:
    """Ask a yes/no question. Defaults to default_yes if the user just presses enter."""
    suffix = "[Y/n]" if default_yes else "[y/N]"
    answer = ask(f"{question} {suffix} ").lower()
    if answer == "":
        return default_yes
    return answer in ("y", "yes")


def ask_repair_reset_abort() -> RepairResetAbortChoice:
    """Ask the repair / reset / abort question used by partial-state handling."""
    answer = ask("Choose an option: [r]epair / reset ([w]ipe and start clean) / [a]bort: ").lower()
    if answer == "repair" or (answer.startswith("r") and not answer.startswith("re")):
        return "repair"
    if answer.startswith("w") or answer == "reset":
        return "reset"
    return "abort"


def is_interactive_terminal() -> bool:
    return sys.stdin.isatty() and sys.stdout.isatty()
